### PR TITLE
Feature: Make executors and feedbacks easier to use outside of the fuzzing loop

### PR DESCRIPTION
**Pull request by Lukas-Dresel**
It would be nice to be able to use LibAFL components more easily outside of LibAFL's fuzzer harnessing (by which I mean the plug-and-play fuzzing loop) to allow for more reuse. 
For example, I found myself ripping my hair out over trying to just run the forkserver with an observer to simply collect, aggregate and plot coverage. To do so, I had to provide types for fuzzers, feedbacks, schedulers, etc. even though none of those things were relevant for my use-case.

I ended up making changes like the ones in this PR in my own fork and it made it quite a bit nicer to use, now you can reuse some of the components and are only required to satisfy the directly needed type constraints for these components.

### CODE REVIEW

**Comment by domenukk** [comment]
@addisoncrump can you take a look? It might interfere with #2438 somewhat (?)

**Comment by domenukk** [comment]
> It would be nice to be able to use LibAFL components more easily outside of LibAFL's fuzzer harnessing (by which I mean the plug-and-play fuzzing loop) to allow for more reuse. For example, I found myself ripping my hair out over trying to just run the forkserver with an observer to simply collect, aggregate and plot coverage. To do so, I had to provide types for fuzzers, feedbacks, schedulers, etc. even though none of those things were relevant for my use-case.

*ideally* the normal run_target would already be reusable / have minimal generic bounds/ have sensible _Nop_ Variants for easy use.
But it might be a goal that's harder to achieve :D

**Comment by addisoncrump** [comment]
Sorry for the delay -- lots of IRL stuff right now. I will look at this tomorrow.

**Comment by domenukk** [comment]
FYI @Lukas-Dresel this is waiting for #2438 which might take a while...
Sorry :)

**Comment by domenukk** [comment]
Sorry for the delay, I think we can merge this.

**Comment by tokatoka** [comment]
wait, i'll check this when i refactor (or before) executor.
don't merge now

**Comment by Lukas-Dresel** [comment]
Just checking in, what's the status of this?

File: libafl/src/feedbacks/stdio.rs [hunk 1]
```diff
@@ -35,6 +36,34 @@ pub struct StdOutToMetadataFeedback {
...
+impl StdOutToMetadataFeedback {
+    /// Append to the testcase the generated metadata in case of a new corpus item.
+    #[inline]
+    fn append_stdout_observation_to_testcase<S, OT>(
+        &mut self,
+        observers: &OT,
+        testcase: &mut Testcase<S::Input>,
+    ) -> Result<(), Error>
+    where
+        S: State + UsesInput,
```
Comment (tokatoka) [inline libafl/src/feedbacks/stdio.rs:48]: Can you try to not use UsesInput here?
You can specify
S: HasCorpus and make the input to be <S::Corpus as Corpus>::Input

File: libafl/src/feedbacks/new_hash_feedback.rs [hunk 2]
```diff
@@ -141,6 +129,36 @@ where
...
+        );
+        Ok(())
+    }
+
+    #[allow(clippy::wrong_self_convention)]
+    fn is_interesting<EM, OT>(
+        &mut self,
+        state: &mut S,
+        _manager: &mut EM,
+        _input: &<S as UsesInput>::Input,
```
Comment (tokatoka) [inline libafl/src/feedbacks/new_hash_feedback.rs:152]: here avoid UsesInput too

File: libafl/src/feedbacks/nautilus.rs [hunk 3]
```diff
@@ -71,6 +71,23 @@ impl<'a, S> NautilusFeedback<'a, S> {
             phantom: PhantomData,
         }
     }
+    fn append_nautilus_metadata_to_state(
+        &mut self,
+        state: &mut S,
+        testcase: &mut Testcase<S::Input>,
+    ) -> Result<(), Error>
+    where
+        S: UsesInput + HasCorpus<Input = NautilusInput> + HasMetadata,
```
Comment (tokatoka) [inline libafl/src/feedbacks/nautilus.rs:80]: here too

**Comment by tokatoka** [comment]
Sorry for the late review.
Overall good. Just one thing, we're trying to remove UsesInput generics recently. Can you try to remove the uses of these trait as much as possible.

File: libafl/src/feedbacks/concolic.rs [hunk 4]
```diff
@@ -44,6 +44,22 @@ impl<'map, S> ConcolicFeedback<'map, S> {
...
         }
     }
+
+    fn add_concolic_feedback_to_metadata<OT>(
+        &mut self,
+        observers: &OT,
+        testcase: &mut Testcase<S::Input>,
+    ) where
+        OT: ObserversTuple<S>,
+        S: UsesInput,
```
Comment (tokatoka) [inline libafl/src/feedbacks/concolic.rs:54]: here too

File: libafl/src/feedbacks/list.rs [hunk 5]
```diff
@@ -119,7 +109,44 @@ where
...
+
+        for v in &self.novelty {
+            history_set.set.insert(*v);
+        }
+    }
+}
+
+impl<S, T> Feedback<S> for ListFeedback<T>
+where
+    S: State + HasNamedMetadata,
```
Comment (tokatoka) [inline libafl/src/feedbacks/list.rs:128]: do you need S: State?

File: libafl/src/feedbacks/map.rs [hunk 6]
```diff
@@ -441,16 +441,16 @@ where
     fn is_interesting<EM, OT>(
         &mut self,
         state: &mut S,
-        manager: &mut EM,
-        input: &<S as UsesInput>::Input,
+        _manager: &mut EM,
+        _input: &<S as UsesInput>::Input,
```
Comment (tokatoka) [inline libafl/src/feedbacks/map.rs:445]: here too.

**Comment by tokatoka** [comment]
tell me if you want me to merge from main and resolve conflicts

**Comment by domenukk** [comment]
@tokatoka if you find the time let's just finish/merge this for 0.14, lukas has done his part I think :)

**Comment by domenukk** [comment]
Finishing this in #2637, thanks for your help @Lukas-Dresel ! :)

**Pull request closed without merging**